versatile-data-kit: VEP-1416 vdk-troubleshooting-tools (#1423)
vmware/versatile-data-kit: specs/vep-1416-vdk-troubleshooting-tools/README.md

  - [API design](#api-design)
  - [Detailed design](#detailed-design)

## Summary

-----
This VEP outlines the changes that would need to be introduced to provide users and Versatile Data Kit administrators with tools to troubleshoot data jobs, which are deployed in a kubernetes cluster. These tools will be provided through a dedicated plugin, which will be configurable and extendable, so more functionality could be added in the future.

The initial troubleshooting capability introduced, is the ability to do a thread dump of the python process and send it to a specific endpoint.

## Glossary

-----
- VDK: https://github.com/vmware/versatile-data-kit/wiki/dictionary#vdk
- Plugins: https://github.com/vmware/versatile-data-kit/wiki/dictionary#vdk-plugins
- Data Job: https://github.com/vmware/versatile-data-kit/wiki/dictionary#data-job
- Data Job Execution: https://github.com/vmware/versatile-data-kit/wiki/dictionary#data-job-execution
- Data Job Deployment: https://github.com/vmware/versatile-data-kit/wiki/dictionary#data-job-deployment
- Kubernetes: https://kubernetes.io/
- Prometheus: https://prometheus.io/docs/introduction/overview/

## Motivation

-----

> When a data job, which is deployed in a kubernetes cluster, fails with an error, the VDK Deployment administrator can troubleshoot it either through the exposed Prometheus metrics or by looking at the job's logs. This works in most cases, but fails when there are issues with the logging service, some dependency does not propagate its logs, or when the configuration of the environment has been changed.

Reviewer: I think another problem that is hard to troubleshoot is if the job is stuck - be it due to a bug in VDK or in the user code - the job hangs .And if it runs with limited OS permissions, it may not be possible to generate even stack trace/thread dump

Author: Good point, note taken. When I have a working implementation, I'll test it with different container permissions and see what could be done to address issues with stuck jobs.

To allow for easier troubleshooting of errors with data jobs, a special plugin will be introduced, which will provide capabilities to do a thread dump and either log the contents or stored locally, where it can be examined by the owner of the data job or by the team responsible for the VDK deployment. The plugin could also be extended in the future with other debugging capabilities.

## Requirements and goals

-----
### Goals

* **Introduce a vdk-jobs-troubleshooting plugin**

* **Provide capability to do a thread dump**
  - A user/admin wants to get a thread dump of the data job process for troubleshooting purposes. For example, a user deploys a data job and after some time the job starts failing without any apparent reason. The user needs to be able to get a thread dump, so they are able to properly investigate the root cause for the issue.

### Non-Goals

> * **Additional troubleshooting tools are not planned as part of this proposal.**

Reviewer: Based on security section - the idea is for this feature to be exposed only for administrators of the VDK deployment and not end users (data engineers - data job developers). 
Which is fine . It makes the design simpler since security implications are different. 

But let's put it as non goal explicitly. My first thought was that end users would be able to use it.

Author: Well, if the data is send to the logging service that the VDK deployment uses, then end-users would, in theory, be able to see the logs (depending on what permissions they have to view logs).

Author: Updated the non-goals section. Thanks

To allow for easier troubleshooting of errors with data jobs, a special plugin will be introduced, which will provide capabilities to do a thread dump and either log the contents or stored locally, where it can be examined by the owner of the data job or by the team responsible for the VDK deployment. The plugin could also be extended in the future with other debugging capabilities.

## Requirements and goals

-----
### Goals

* **Introduce a vdk-jobs-troubleshooting plugin**

* **Provide capability to do a thread dump**
  - An administrator wants to get a thread dump of the data job process for troubleshooting purposes. For example, a user deploys a data job and after some time the job starts failing without any apparent reason. The administrator needs to be able to get a thread dump, so they are able to properly investigate the root cause for the issue.

### Non-Goals

* **Additional troubleshooting tools are not planned as part of this proposal.**
* **The thread-dump feature would not be exposed to users/data job owners.**

## High-level design

-----

![plugin_diagram.png](diagrams/plugin_diagram.png)

> 

Reviewer 2: Missing the new components descriptions (I see a new plugin, and a new server).

Author: Will add then in a subsequent PR. Thanks for the review

For the proposed design, a vdk-jobs-troubleshooting plugin will be introduced. The plugin will start a http server instance, which will allow users to obtain a thread dump (which will also be printed in the logs) for further troubleshooting.

> 

Reviewer 2: what is "invoke", can you elaborate on what the invocation is?

Author: Note taken, will provide more details in subsequent PR.

## API design

-----

> No changes to the public API.

Reviewer 2: This section describing the newly added HTTP server API

Author: Note taken, will elaborate in a subsequent PR.

> 

Reviewer 2: Is there any change on the k8s pod interface surface? Where is the HTTP server listening / awaiting requests? What is a valid request?

Author: The HTTP server will be listening on localhost, so no port or anything will be exposed to the outside world. The connection will happen when an administrator with access to the k8s cluster does a port-forward to the data job pod.
Will provide more details in a subsequent PR.
Thank you for the review

## Detailed design

-----

### vdk-jobs-troubleshooting Plugin
The plugin will act as a toolbox, where data job troubleshooting tools will be implemented. As part of this proposal, only a thread-dump utility will be implemented. It will start a local web server, which will attach to port 8080 and, depending on configuration, will allow for the dump to be logged as part of normal logging processes, or a user can do port-forwarding to the data job pod and examine the contents of the thread data.

![plugin_details.png](diagrams/plugin_details.png)

The main configuration variable for this plugin would be:
* VDK_TROUBLESHOOT_UTILITIES_TO_USE, which will accept a comma-separated list of string literals with the troubleshooting utilities that will be used, e.g., `"utility1,utility2,..."`.

### Availability

> The plugin will be part of the vdk installation, so the same availability constraints apply. The http server will be running in a separate process, but it will still be available as long as the data job pod is running. In case of issues with the logging service, the http server would still be able to log the thread-dump locally in the data job pod.

Reviewer 2: is the HTTP server a separate process? Why is it visualised within the VDK box, what is the VDK box?

Author: The VDK box refers to the VDK container in the Data Job kubernetes pod. The http server is a separate process, but it is confined within the container. I realise that the diagram is kind of misleading, in the sense that the vdk container is an init container, and the "main" container is the data job one, so will rework the diagram and provide more explanations in a subsequent PR.

# VEP-1416: VDK Troubleshooting Tools

* **Author(s):** Andon Andonov (andonova@vmware.com)
* **Status:** draft

  - [Summary](#summary)
  - [Glossary](#glossary)
  - [Motivation](#motivation)
  - [Requirements and goals](#requirements-and-goals)
  - [High-level design](#high-level-design)
  - [API design](#api-design)
  - [Detailed design](#detailed-design)

## Summary
This VEP outlines the changes that would need to be introduced to provide users and Versatile Data Kit administrators with tools to troubleshoot data jobs, which are deployed in a kubernetes cluster. These tools will be provided through a dedicated plugin, which will be configurable and extendable, so more functionality could be added in the future.

> The initial troubleshooting capability that will be introduced, will be the ability to do a thread dump of the python process and send it to a specific endpoint.

Reviewer 2: ```suggestion
The initial troubleshooting capability introduced, is the ability to do a thread dump of the python process and send it to a specific endpoint.
```

To allow for easier troubleshooting of errors with data jobs, a special plugin will be introduced, which will provide capabilities to do a thread dump and send the data to a predefined endpoint where it can be examined by the owner of the data job or by the team responsible for the VDK deployment. The plugin could also be extended in the future with other debugging capabilities.

## Requirements and goals
### Goals

* **Introduce a vdk-jobs-troubleshooting plugin**

* **Provide capability to do a thread dump**
  - A user/admin wants to get a thread dump of the data job process for troubleshooting purposes. For example, a user deploys a data job and after some time the job starts failing without any apparent reason. The user needs to be able to get a thread dump, so they are able to properly investigate the root cause for the issue.

> 

Reviewer 2: Are there any prerequisites on the security context?

Author: Check the `Security and Permissions` section.

### Non-Goals

* **Additional troubleshooting tools are not planned as part of this proposal.**

## High-level design

<!--
All the rest sections tell **how** are we solving it?

This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?

Provide a valid UML Component diagram that focuses on the architecture changes
implementing the feature. For more details on how to write UML Component Spec -
see https://en.wikipedia.org/wiki/Component_diagram#External_links.

For every new component on the diagram, explain which goals does it solve.
In this context, a component is any separate software process.

-->
![plugin_diagram.png](plugin_diagram.png)

> For the proposed design, a vdk-jobs-troubleshooting plugin will be introduced. The plugin will start a http server instance, which will monitor the python process of the execution, and in case of an error with the job, will do a thread dump and send it to a predefined endpoint.

Reviewer 3: ```suggestion
For the proposed design, a vdk-jobs-troubleshooting plugin will be introduced. The plugin will start a http server instance, which will allow users to obtain a thread dump (which will also be printed in the logs) for further troubleshooting.
```
